ada_feeding_perception/ada_feeding_perception/table_detection.py

```diff
@@ -0,0 +1,483 @@
+        depth_mask = cv2.morphologyEx(depth_mask, cv2.MORPH_CLOSE, kernel, iterations=3)
+
+        # Remove Outliers
+        # TODO: Consider outlier removal using median + iqr for the future
+        depth_var = np.abs(depth_mask - np.mean(depth_mask[depth_mask > 0]))
+        depth_std = np.std(depth_mask[depth_mask > 0])
+        depth_mask[depth_var > 2.0 * depth_std] = 0
+
+        # Fit Plane: depth = a*u + b*v + c
+        d_idx = np.where(depth_mask > 0)
+        d = depth_mask[d_idx].astype(float)
+        coeffs = np.hstack((np.vstack(d_idx).T, np.ones((len(d_idx[0]), 1)))).astype(
+            float
+        )
+
+        # Coefficients b and a are reversed because of matrix row/col structure and its
+        # correspondence to x/y
+        b, a, c = np.linalg.lstsq(coeffs, d, rcond=None)[0]
+
+        # Create Table Depth Image
+        u = np.linspace(0, depth_mask.shape[1], depth_mask.shape[1], False)
+        v = np.linspace(0, depth_mask.shape[0], depth_mask.shape[0], False)
+        u_grid, v_grid = np.meshgrid(u, v)
+        table = a * u_grid + b * v_grid + c
+        table = table.astype("uint16")
+
+        return table
+
+    def get_orientation(
+        self,
+        camera_info: CameraInfo,
+        table_depth: Image,
+    ):
+        """
+        Calculate the orientation of the table plane with respect to the
+        camera's frame of perspective.
+
+        Parameters
+        ----------
+        camera_info: The camera information.
+        table_depth: The depth image of the table.
+
+        Returns
+        ----------
+        center: The center coordinates of the table in the camera frame.
+        q: The quaternion representing the orientation of the table plane.
+        """
+        # Deproject the depth image to get the 3D points in the camera frame.
+        pointcloud = depth_img_to_pointcloud(
+            table_depth,
+            0,
+            0,
+            f_x=camera_info.k[0],
+            f_y=camera_info.k[4],
+            c_x=camera_info.k[2],
+            c_y=camera_info.k[5],
+        )
+        xy_d = np.concatenate(
+            (pointcloud[:, 0, np.newaxis], pointcloud[:, 1, np.newaxis]), axis=1
+        )
+        depth_d = pointcloud[:, 2]
+        center = [
+            pointcloud[self.deproject_center_coord][0],
+            pointcloud[self.deproject_center_coord][1],
+            pointcloud[self.deproject_center_coord][2],
+        ]
+
+        # Fit Plane: z = a*x + b*y + c
+        coeffs = np.hstack((np.vstack(xy_d), np.ones((len(xy_d), 1)))).astype(float)
+        a, b, c = np.linalg.lstsq(coeffs, depth_d, rcond=None)[0]
+
+        # Modify the z coordinate of the center given the fitted plane (i.e. make the center
+        # the origin of the table plane)
+        center[2] = a * center[0] + b * center[1] + c
+
+        # Get the direction vectors of the table plane from the camera's frame of perspective
```

Comment:
I understand what you're doing because we discussed it thoroughly, but by reading the code itself I was very confused. This has to do with several issues:
1. Unclear variable names -- for example, what is `x_p`? What is `denom`?
2. Comments that don't align with the code. You write `direction vector` in the commend, but it took me a while to figure out that those correspond to `x_ref, y_ref, z_ref`.

Please revamp this part to improve readability.